adcm_aio_client/core/requesters.py

```diff
@@ -22,20 +40,129 @@ def as_dict(self: Self) -> dict: ...
 
 
 class Requester(Protocol):
-    async def get(self: Self, path: str, query_params: dict) -> AsyncGenerator[RequesterResponse]: ...
+    async def get(self: Self, path: str, query_params: dict) -> RequesterResponse: ...
+
+    async def post(self: Self, path: str, data: dict) -> RequesterResponse: ...
+
+    async def patch(self: Self, path: str, data: dict) -> RequesterResponse: ...
+
+    async def delete(self: Self, path: str) -> RequesterResponse: ...
+
+
+@dataclass(slots=True, frozen=True)
+class Credentials:
+    username: str
+    password: str
+
+    def dict(self: Self) -> dict:
+        return asdict(self)
+
+    def __repr__(self: Self) -> str:
+        return f"{self.username}'s credentials"
+
+
+@dataclass(slots=True, frozen=True)
+class HTTPXRequesterResponse:
+    response: httpx.Response
+
+    def as_list(self: Self) -> list:
+        if not isinstance(data := self._prepare_json_response(), list):
+            message = f"Expected a list, got {type(data)}"
+            raise RequesterResponseError(message)
+
+        return data
+
+    def as_dict(self: Self) -> dict:
+        if not isinstance(data := self._prepare_json_response(), dict):
+            message = f"Expected a dict, got {type(data)}"
+            raise RequesterResponseError(message)
+
+        return data
 
-    async def post(self: Self, path: str, data: dict) -> AsyncGenerator[RequesterResponse]: ...
+    def _prepare_json_response(self: Self) -> Json:
+        try:
+            data = self.response.json()
+        except JSONDecodeError as e:
+            message = "Response can't be parsed to json"
+            raise RequesterResponseError(message) from e
 
-    async def patch(self: Self, path: str, data: dict) -> AsyncGenerator[RequesterResponse]: ...
+        if "results" in data:
+            data = data["results"]
 
-    async def delete(self: Self, path: str) -> AsyncGenerator[RequesterResponse]: ...
+        return data
 
 
-class Session: ...
+def _get_error_class(status_code: int) -> type[ResponseError]:
+    first_digit = status_code // 100
+    match first_digit:
+        case 5:
+            return ServerError
+        case 4:
+            match status_code:
+                case 400:
+                    return BadRequestError
+                case 401:
+                    return UnauthorizedError
+                case 404:
+                    return NotFoundError
+                case 409:
+                    return ConflictError
+
+    return ResponseError
+
+
+def convert_exceptions(func: Callable) -> Callable:
+    async def wrapper(*arg: Iterable, **kwargs: dict) -> httpx.Response:
+        response = await func(*arg, **kwargs)
+        try:
+            response.raise_for_status()
+        except httpx.HTTPStatusError as e:
+            raise _get_error_class(status_code=response.status_code) from e
```

Comment:
raising and catching in this case introduces unnecessary overhead, I believe

 let's just check status code for >= 300 (300 is not an error, but for us it is, since we haven't used redirect)

Reply:
fixed